### ISSUE TITLE
TLS FFI API

### ISSUE THREAD

**Issue by winterland1989** [randombit/botan#2492]
Just to make sure, is there any plan to add TLS APIs to ffi.h ? I'm trying add botan binding to Haskell and I'm not sure if I sure add wrappers around current C++ APIs.

**Comment by randombit** [comment]
This is definitely the plan in the long term, so it can also be used from Rust/Python/etc, but it's not likely to happen in say the next 6 months.

**Comment by bitberlin** [comment]
TLS support in FFI API would be very fine! This is currently the only reason why i can't use botan for my project.

**Comment by benrushcc** [comment]
Hi, I'm checking in on this old issue about exposing TLS APIs via the FFI.

I noticed that there is a refs/heads/jack/ffi-tls branch in the repository, which seems to suggest some related work might be in progress. Is that branch part of the plan to address this?

I'm currently evaluating Botan for a project that needs TLS support through C bindings (for Java). Knowing the current status and roadmap would be very helpful for my planning.

Thanks!

**Comment by randombit** [comment]
@benrushcc That particular branch is not really complete or ready to go in any way. There is an effort currently underway to get funding to support the completion of this significant project. We'll probably know more in the next few months. But even if funding comes through it likely will not be available until - at the absolute most optimistic - sometime in 2027.

**Comment by moritzschmitt** [comment]
@randombit We are interested in this issue because we would like to build a Go library that uses Botan for TLS through the C API. We have some experience with the FFI already (most recently #5903).

Before we invest time, I would like to ask whether a community contribution is wanted here, or whether it would conflict with the funding effort you mentioned above. We do not want to get in the way of a funded project.

If a contribution is welcome, we would start with a proposal rather than code. This would be a draft of the header and the reasoning behind the main decisions (I/O model, credentials, session handling, error reporting), meant as the starting point for a design discussion. So far we have only a rough idea of how this could work. We can post it here or as a GitHub Discussion, whichever you prefer.

**Comment by randombit** [comment]
@moritzschmitt Just got word that the funding we had hoped for which would have covered FFI-TLS along with some other work will not occur (alas), so at this point a contribution here would be completely welcome. Let's use this issue for design discussions.

**Comment by moritzschmitt** [comment]
@randombit Sorry to hear about the funding, but we will surely be able to do without.

Here is a first sketch of how we could approach it. The idea is a thin C shim over the existing C++ TLS API, without any new protocol code. Please read it critically. Our understanding of the TLS code is new and still brittle, so we may well have got details wrong.

* Layout as in `jack/ffi-tls`. A separate `ffi_tls.h` in a new `src/lib/ffi/ffi_tls` submodule that requires `tls`, with its own `FFI_TLS` version stamp and `botan_ffi_tls_api_version()` in `ffi.h`, so applications can detect the module at run time.
* One opaque handle per collaborator. `botan_tls_policy_t`, `botan_tls_credentials_t`, `botan_tls_session_manager_t` and `botan_tls_channel_t`. Since `TLS::Client` and `TLS::Server` take their collaborators as `shared_ptr`, the first three handles would wrap a `shared_ptr` (as `ffi_tpm2.cpp` does). A channel then keeps them alive, and the application may destroy the handles right after creating the channel.
* I/O through callbacks, mirroring `TLS::Callbacks`. The application feeds ciphertext with `botan_tls_channel_received_data()` and sends plaintext with `botan_tls_channel_send()`. Four C callbacks correspond to `tls_emit_data`, `tls_record_received`, `tls_alert` and `tls_session_activated`. They take a `void*` context first and return `int`, like the callbacks of `botan_rng_init_custom`, so an application can abort the connection by returning nonzero. Certificate verification stays with the default `tls_verify_cert_chain`.
* Policies by name (`Policy`, `Strict_Policy`, `BSI_TR_02102_2`, `NSA_Suite_B_128`, `NSA_Suite_B_192`) plus `Text_Policy` from a string, so no per-setting functions are needed.
* Credentials as a small setter based object backed by an internal `Credentials_Manager` (trusted certificates, CRLs, certificate chains with private keys). Inputs are copied, so the certificate and key handles need not outlive the object.
* Session managers limited to `Session_Manager_In_Memory` and `Session_Manager_Noop`.
* Errors reuse the existing codes, mainly `BOTAN_FFI_ERROR_TLS_ERROR`, plus a getter for the last alert, because the alert code is otherwise lost at the FFI boundary.

We would deliberately leave DTLS, PSK, a verification callback, session export, key material export and custom crypto hooks out of the first version. Each of them can be added later as a new function without changing the existing ones.

As a first step we could open a draft PR with the header, the module skeleton and a loopback test in `src/tests/test_ffi_tls.cpp` modelled on `unit_tls.cpp`, so that names and lifetimes can be reviewed early. The Python binding would follow in the same PR.

**Comment by moritzschmitt** [comment]
And one more question: Would you accept `lifecycle -> "Experimental"` on the new module for its first release, so that signatures may still change after some real world use?

**Comment by randombit** [comment]
Outline looks fine to me and more or less how I had planned it.

> Would you accept lifecycle -> "Experimental" on the new module for its first release

Yes that is quite reasonable and makes early mistakes less costly.

> Policies by name (Policy, Strict_Policy, BSI_TR_02102_2, NSA_Suite_B_128, NSA_Suite_B_192) plus Text_Policy from a string, so no per-setting functions are needed.

I think we would eventually want to support a full override of the policies, but an opaque policy type that initially just has initializers with our stock policies is fine to start. [Skip the suite B policies though, IIRC suite B has been entirely withdrawn]

**Comment by reneme** [comment]
> `botan_tls_policy_t`, `botan_tls_credentials_t`, `botan_tls_session_manager_t` and `botan_tls_channel_t`. Since `TLS::Client` and `TLS::Server` take their collaborators as `shared_ptr`, the first three handles would wrap a `shared_ptr` (as `ffi_tpm2.cpp` does). A channel then keeps them alive, and the application may destroy the handles right after creating the channel.

It is worth noting that the helpers may be used across multiple instances of TLS channels. Internally, that lifetime is still handled by the shared pointers, but it might have implications on the FFI API design that aren't immediately obvious.

> Four C callbacks correspond to `tls_emit_data`, `tls_record_received`, `tls_alert` and `tls_session_activated`.

To this list I would add `tls_session_established` which is similar to `tls_session_activated` but it provides versatile information about the established session's parameters via [the `Session_Summary` struct](https://github.com/randombit/botan/blob/85207728a75aeedf669655f5874fbef0c2fa11b4/src/lib/tls/tls_session.h#L147) that should be (partially) introspectable via new FFI API. This includes information about the negotiated crypto parameters, whether or not session resumption was used and, perhaps most importantly, the received peer certificate.

> We would deliberately leave DTLS, PSK, a verification callback, session export, key material export and custom crypto hooks out of the first version. Each of them can be added later as a new function without changing the existing ones.

That's very much sensible as each of them add complexity to the initial API surface unnecessarily. Though, I'd suggest to consider adding a _verification callback_ as soon as possible. It should be fairly easy to implement as just another optional callback (much like `tls_emit_data`, `tls_record_received` and friends) and the required FFI API for handling certificates is available already today. In my experience, most applications are quite opinionated on the certificates involved in the TLS handshake. Some checks might be implementable after the handshake has completed using status information delivered through `tls_session_established`, but being able to abort a handshake as soon as the peer presents its certificate is definitely preferable.

---

Perhaps obvious, but for completeness: we'll very likely need a handful of functions on the `botan_tls_channel_t` handle. On the one hand, that will be simple getters such as `is_active`, `is_handshake_complete`, and `is_closed`. But also functions with side effects such as `close`, or `send_alert`.

**Comment by moritzschmitt** [comment]
@reneme Thanks, that is very helpful.

* Shared helpers: We would document that the setters of a policy, credentials or session manager object must not be called once the object has been handed to a channel, and keep everything that belongs to a single connection (for example a verification callback and its context pointer) on the channel rather than on the shared objects.
* `tls_session_established`: Agreed. We would add it as a callback and make the `Session_Summary` fields available through getters on the channel handle (negotiated version and ciphersuite, resumption, the peer certificates as `botan_x509_cert_t`, and so on).
* Verification callback: Agreed as well, we would include it early. A first version would receive the chain as certificate handles, the usage and the hostname. OCSP responses and the trusted roots have no FFI representation yet, so they would have to come later. We would suggest running the default PKIX validation first and then the application callback for additional checks, with an option to skip the default validation for applications that want full control. Does that sound reasonable to you?
* `send_alert` and the getters you list will be there.

One consequence of the growing list of callbacks is that we would rather not pass them as parameters of the init functions. We are thinking of a `botan_tls_callbacks_t` handle with one setter per callback, so that further callbacks (such as the deferred operation callback from #5915) can be added later without changing existing signatures.

**Comment by reneme** [comment]
> Regarding the Verification Callback: We would suggest running the default PKIX validation first and then the application callback for additional checks, with an option to skip the default validation for applications that want full control. Does that sound reasonable to you?

Yes! I meant to propose that as well, but forgot about it in the end. Running the default verification is a good baseline. Conceptually much like overriding `Callbacks::tls_verify_cert_chain` and up-calling the base implementation first. I like that! But a user should be able to opt-out of it as well.

> `tls_session_established`: We would add it as a callback and make the Session_Summary fields available through getters on the channel handle (negotiated version and ciphersuite, resumption, the peer certificates as botan_x509_cert_t, and so on).

Initially, I thought of a `tls_session_summary_t` handle with getters, frankly. However, just collecting the summary information internally (in the machinery behind the FFI API) and exposing it on the channel handle as you suggested could also be just fine and would avoid the explicit "established" callback altogether. Before the handshake succeeded those generic getters would then just error-out with `BOTAN_FFI_ERROR_INVALID_OBJECT_STATE`, I guess. 

Both options make sense to me, but avoiding an explicit `tls_session_summary_t` handle definitely saves some boilerplate and API surface.

> One consequence of the growing list of callbacks is that we would rather not pass them as parameters of the init functions. We are thinking of a botan_tls_callbacks_t handle with one setter per callback, so that further callbacks (such as the deferred operation callback from https://github.com/randombit/botan/pull/5915) can be added later without changing existing signatures.

That's definitely the most future-proof design because more callbacks can be added later. Maybe let the initializer function of `botan_tls_callbacks_t` take all the mandatory callbacks -- this part of the callbacks-API proofed very stable in many years (apart from `std::span<>` adaptions for C++20). I.e. like so:

```C++
int botan_tls_callbacks_init(botan_tls_callbacks_t *cb, 
                             botan_tls_emit_data_fn edfn,
                             botan_tls_record_received_fn rrfn,
                             botan_tls_alert_received_fn arfn);
```

... that way we ensure "valid on construction" semantics for the callbacks handle.

**Comment by moritzschmitt** [comment]
> Yes! I meant to propose that as well, but forgot about it in the end. Running the default verification is a good baseline. Conceptually much like overriding `Callbacks::tls_verify_cert_chain` and up-calling the base implementation first. I like that! But a user should be able to opt-out of it as well.

Good, then that is settled.

> Initially, I thought of a `tls_session_summary_t` handle with getters, frankly. However, just collecting the summary information internally (in the machinery behind the FFI API) and exposing it on the channel handle as you suggested could also be just fine and would avoid the explicit "established" callback altogether. Before the handshake succeeded those generic getters would then just error-out with `BOTAN_FFI_ERROR_INVALID_OBJECT_STATE`, I guess.

Agreed, getters on the channel handle and no separate summary handle. Before the handshake they would return `BOTAN_FFI_ERROR_INVALID_OBJECT_STATE`, and `BOTAN_FFI_ERROR_NO_VALUE` only once the handshake is done but nothing was negotiated (for example no ALPN protocol). We would still keep a callback for `tls_session_activated`, since it tells the application when it may start sending, and drop the explicit `tls_session_established` one.

> int botan_tls_callbacks_init(botan_tls_callbacks_t *cb, 
>                              botan_tls_emit_data_fn edfn,
>                              botan_tls_record_received_fn rrfn,
>                              botan_tls_alert_received_fn arfn);

Looks good to me. `botan_tls_callbacks_init` would take the three mandatory callbacks, which matches the three pure virtuals of `TLS::Callbacks`. One detail to settle is the application context pointer. We would pass it to the channel init function rather than store it in the callbacks handle, so that one handle can serve any number of connections with different contexts. What do you think?

**Comment by reneme** [comment]
> One detail to settle is the application context pointer. We would pass it to the channel init function rather than store it in the callbacks handle, so that one handle can serve any number of connections with different contexts. What do you think?

My first inclination would be to bind the application context to the channel (i.e. passed as a `void*` into `botan_tls_channel_init`) and then all callback functions (by convention) receive the application context as its first parameter. Thus the callbacks handle (and also the other helper handles) would not maintain an application context pointer themselves, but always get passed the application context associated with the concrete channel instance.

As I said earlier: all these helpers may be shared between multiple channels. Technically, that is true even for the `Callbacks`, but often that's not really feasible for other practical reasons.

Frankly, we'll probably have to iterate on such details during the implementation. Not sure what I wrote above will hold once there's actual code being produced.

**Comment by moritzschmitt** [comment]
The first PR is up as #5941 (draft). It is smaller than the loopback test I announced above: only the `ffi_tls` module skeleton with the experimental lifecycle, the two runtime version functions in `ffi.h`, and the `botan_tls_policy_t` handle with the stock and text policies, plus tests, the Python binding and docs. The idea is to get the conventions (module layout, experimental status, handle pattern, naming, test and docs placement) reviewed before the credentials, session manager and channel parts build on them. Feedback on any of that is very welcome, there or here.
